File: python/652_寻找重复的子树_medium.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def findDuplicateSubtrees(self, root: TreeNode) -> List[TreeNode]:
        treeMap = collections.Counter()
        ans = []
        def deepTraverse(root):
            if not root:
                return '#'
            seq = '{},{},{}'.format(root.val, deepTraverse(root.left), deepTraverse(root.right))
            treeMap[seq] += 1
            if treeMap[seq] == 2:
                ans.append(root)
            return seq
        deepTraverse(root)
        return ans

File: cpp/652_寻找重复的子树_medium.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<TreeNode*> findDuplicateSubtrees(TreeNode* root) {
        unordered_map <string, int> treeMap;
        vector<TreeNode *> ans;
        deepTraverse(treeMap, root, ans);
        return ans;
        
    }
    string deepTraverse(unordered_map<string, int> &treeMap, TreeNode* &root, vector<TreeNode*> &ans){
        if(!root)
            return "#";
        string seq = char(root->val) + deepTraverse(treeMap,root->left,ans) + deepTraverse(treeMap,root->right,ans);
        treeMap[seq]++;
        if(treeMap[seq]==2)
            ans.push_back(root);
        return seq;
    }

};